tidy3d/components/medium.py

```diff
@@ -82,24 +82,122 @@ def _eps_model(self, frequency: float) -> complex:
 """ Medium Definitions """
 
 
-class NonlinearSpec(ABC, Tidy3dBaseModel):
-    """Abstract specification for adding a nonlinearity to a medium.
+class NonlinearModel(ABC, Tidy3dBaseModel):
+    """Abstract model for a nonlinear material response.
+    Used as part of a :class:`.NonlinearSpec`."""
+
+    def _validate_medium_type(self, medium: AbstractMedium):
+        """Check that the model is compatible with the medium."""
+        if isinstance(medium, AbstractCustomMedium):
+            raise ValidationError(
+                f"'NonlinearModel' of class {type(self)} is not currently supported "
+                f"for medium class {type(medium)}."
+            )
+        if not isinstance(medium, (Medium, DispersiveMedium)):
+            raise ValidationError(
+                f"'NonlinearModel' of class {type(self)} is not currently supported "
+                f"for medium class {type(medium)}."
+            )
+
+    def _validate_medium(self, medium: AbstractMedium):
+        """Any additional validation that depends on the medium"""
+        pass
+
+    def _validate_medium_freqs(self, medium: AbstractMedium, freqs: List[pd.PositiveFloat]) -> None:
+        """Any additional validation that depends on the central frequencies of the sources."""
+        pass
+
+    def _get_n0(
```

Comment:
I went through to try to understand this a bit better and refactored a bit, feel free to implement if it makes it clearer.

```py
        freqs = set(freqs)

        if not len(freqs):
            if n0 is None:
                raise ...
            return n0

        if len(freqs) > 1:
            if n0 is None:
                raise ...
            log.warning(...)
            return n0

        if len(freqs) == 1:
            freq = freqs[0]
            n, k = medium.nk_model(freq)
            nk = n + 1j * k

            if n0 is not None and not np.isclose(n0, nk):
                log.warning(...)
            return nk

```

Reply:
thanks, I used some of this in the revision